src/Database/LSMTree/Internal/MergingRun.hs

```diff
@@ -211,265 +227,504 @@ duplicateRuns (DeRef mr) =
+    atomicModifyInt var $ \(CreditsPair (SpentCredits   !spent)
+                                        (UnspentCredits !unspent)) ->
+      let spent'   = spent   + spend
+          unspent' = unspent - spend
+          after    = CreditsPair (SpentCredits   spent')
+                                 (UnspentCredits unspent')
+       in assert (spent' + unspent' == spent + unspent) $
+          (after, ())
+
+{-------------------------------------------------------------------------------
+  The main algorithms
+-------------------------------------------------------------------------------}
 
-{-# SPECIALISE takeAllUnspentCredits ::
-     UnspentCreditsVar RealWorld
+{-# SPECIALISE supplyCredits ::
+     Ref (MergingRun IO h)
+  -> CreditThreshold
+  -> Credits
   -> IO Credits #-}
--- | In a CAS-loop, subtract all unspent credits and return them.
-takeAllUnspentCredits ::
-     PrimMonad m
-  => UnspentCreditsVar (PrimState m)
+-- | Supply the given amount of credits to a merging run. This /may/ cause an
+-- ongoing merge to progress.
+supplyCredits ::
+     forall m h. (MonadSTM m, MonadST m, MonadMVar m, MonadMask m)
+  => Ref (MergingRun m h)
+  -> CreditThreshold
+  -> Credits
   -> m Credits
-takeAllUnspentCredits (UnspentCreditsVar !unspentCreditsVar) = do
-    prev <- readPrimVar unspentCreditsVar
-    casLoop prev
-  where
-    casLoop !prev = do
-      prev' <- casInt unspentCreditsVar prev 0
-      if prev' == prev then
-        pure (Credits prev)
-      else
-        casLoop prev'
-
-{-# SPECIALISE stepMerge ::
+supplyCredits (DeRef MergingRun {
+                 mergeKnownCompleted,
+                 mergeNumEntries,
+                 mergeCreditsVar,
+                 mergeState
+               })
+              !creditBatchThreshold !credits =
+    assert (credits >= 0) $ do
+    mergeCompleted <- readMutVar mergeKnownCompleted
+    case mergeCompleted of
+      MergeKnownCompleted -> pure credits
+      MergeMaybeCompleted -> do
+        -- Atomically add credits to the unspent credits (but not allowing
+        -- 'suppliedCredits' to exceed the total debt), determine which case
+        -- we're in and thus how many credits we should try to spend now on
+        -- performing merge steps. Return the credits to spend now and any
+        -- leftover credits that would exceed the debt limit.
+        (spendCredits,
+         leftoverCredits) <- atomicDepositAndSpendCredits
+                               mergeCreditsVar
+                               (numEntriesToTotalDebt mergeNumEntries)
+                               creditBatchThreshold credits
+
+        when (spendCredits > 0) $ do
+          weFinishedMerge <-
+            performMergeSteps mergeState mergeCreditsVar spendCredits
+              `onException`
+            -- If an exception occurs while merging (sync or async) then we
+            -- reverse the spending of the credits.
+            atomicSpendCredits mergeCreditsVar (-spendCredits)
```

Comment:
I think using `bracketOnError` would make it seem more obviously correct to me, but it should be fine to depend on being in a masked state here.

Also, I think we can leave out the `when (spendCredits > 0)` branch, since it's handled by `performMergeSteps`. But again, it's okay the way it is.

Reply:
It's an interesting point that we're almost certainly doing a lot more mask manipulation than we need to. We could probably just mask by default everywhere and rely on synchronous interruption points to ensure prompt delivery of async exceptions.